Sources/IntegerUtilities/GCD.swift

```diff
@@ -0,0 +1,47 @@
+//===--- GCD.swift --------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Numerics open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift Numerics project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension BinaryInteger {
+  
+  /// The greatest common divisor of `a` and `b`.
+  ///
+  /// If both inputs are zero, the result is zero. If one input is zero, the
+  /// result is the absolute value of the other input.
+  ///
+  /// [wiki]: https://en.wikipedia.org/wiki/Greatest_common_divisor
+  @inlinable
+  public static func gcd(_ a: Self, _ b: Self) -> Self {
+    var x = a.magnitude
+    var y = b.magnitude
+    
+    if x == 0 { return Self(y) }
+    if y == 0 { return Self(x) }
+    
+    let xtz = x.trailingZeroBitCount
+    let ytz = y.trailingZeroBitCount
+    
+    x >>= xtz
+    y >>= ytz
+    
+    // The binary GCD algorithm
+    //
+    // At the top of the loop both x and y are odd. Each pass removes at least
+    // one low-order bit from the larger of the two, so the number of iterations
+    // is bounded by the sum of the bit-widths of the inputs.
+    while x != 0 {
+      if x < y { swap(&x, &y) }
+      x -= y
+      x >>= x.trailingZeroBitCount
+    }
```

Comment:
Thinking about micro-optimizations, should the loop be rearranged like this?

```
repeat {
  x >>= x.trailingZeroBitCount
  if x < y { swap(&x, &y) }
  x -= y
} while x != 0
```

It saves one `x != 0` (on the first pass), and allows removing the earlier line `x >>= xtz` to save one shift operation (on the last pass when `x` is zero).

The only drawback I see is that the explanatory comment would need to specify a location other than the top of the loop for where both values are odd, but that’s not really a problem.

Reply:
I am not really convinced that this is even an optimization; the compiler can basically do this for you, because x == 0 has already been checked above (so the compiler knows statically that x != 0 and will skip the first check), and people are generally more familiar with while than repeat-while, so there's some advantage. You can do _half_ of the change and eliminate the shift, I guess:

```
while x != 0 {
  x >>= x.trailingZeroBitCount
  if x < y { swap(&x, &y) }
  x -= y
}
```

Reply:
True, though I must say it’s not at all obvious “how smart” the compiler is, or what optimizations are guaranteed to occur. I think for now I’m going to leave the simple, symmetric version of the code as-is. If benchmarks reveal ways to make it more efficient, that can always be done later.

Another potential optimization would be, if either `x` or `y` equals 1 after the initial right-shifts, then we can just return `1 << min(xtz, ytz)`. Of course that check might slow things down the rest of the time, so it’s another place for benchmarking. I think we can safely leave the micro-optimizations for the future.

Reply:
The optimization in question here is absolutely trivial and does not require any real compiler smarts.

Reply:
All right, I made `gcd` a free function, and then (in a separate commit for ease of diff-viewing) moved the right-shift to the top of the loop and eliminated the two earlier lines that are consequently unnecessary.